### ISSUE TITLE
Question about benchmark or smoothness

### ISSUE THREAD

**Issue by wontheone1** [oakes/play-cljs#19]
Hi again, hopefully it doesn't feel like I am spamming the repo.

When I ran super-koalio example, the game render is not very smooth and felt like fps is under 60.
I like the approach of this game engine as I think it's more declarative and data driven. Render according to game state is I think similar idea as to popular React Js. (Is this your intention? I think it will be nice to write your philosophy about game engine to README to be more clear of your intention.)
But I don't know if benchmark and performance is taken a toll because of this approach, do you think the performance issue is real (or is it just me who feels like it's slow)? And if yes, do you know how to solve it? @oakes

**Comment by oakes** [comment]
I tried both firefox and safari, and they are far smoother. I checked with older versions of play-cljs and clojurescript, and noticed the same difference. For example, [this](https://oakes.github.io/functional-game-dev/) was compiled with play-cljs 0.10.0 and CLJS 1.9.456, and it too is slow in chrome and fast in firefox/safari. I am inclined to believe that chrome's performance has degraded for some reason. Very weird.

In chrome dev tools, you can see it's spending a lot of time on basic canvas operations (getImageData, drawImage) which are called by the tiled map renderer:

![image](https://user-images.githubusercontent.com/216147/42162376-741380c8-7dcc-11e8-81de-ab57e62c4c66.png)

**Comment by wontheone1** [comment]
@oakes Yes I tried in safari and it feels much faster / smoother than from chrome

**Comment by wontheone1** [comment]
But this slow problem also worsens having to write custom physics (https://github.com/oakes/play-cljs/issues/16),
When frame is low and therefore delta time is high, Koalio fall down too much in one frame so it does not get detected for collision on the ground and just falls through below it.

**Comment by wontheone1** [comment]
if this game engine can do some optimizations like the ReactJS (compare the difference in memory and render only the difference), it would be great.

**Comment by oakes** [comment]
Using React would be interesting, but a very big change. I actually think React would work better with SVG rather than play-cljs' canvas-based rendering. SVG works inside the DOM, which is what React is meant for.

In the near term, I'd like to track down why chrome is so much slower. I haven't found any recent bug reports about this yet. Maybe it is something very subtle that doesn't happen often in canvas-based games. Let me know if you find anything.

**Comment by Toad-Racer** [comment]
I don't have much to add, just that I was experiencing this issue more than a year ago. So it's not a recent chrome bug or anything.

**Comment by oakes** [comment]
The weird part is, I used [this demo game](https://oakes.github.io/functional-game-dev/) to give a talk to college students a few years ago, and I ran it in chrome with no issues. In fact, [here's a video of the talk](https://www.youtube.com/watch?v=vFbaUhCYO3Y), lol... You can tell the game ran much smoother back then. So I'm leaning towards concluding that something happened to chrome in the meantime.

**Comment by markx** [comment]
I haven't tested, but I heard p5 is not famous for its performance for gamedev. I wished you had used pixijs instead.  :P

There are people building other renderers for React, like https://github.com/Izzimach/react-pixi or even  https://github.com/weavejester/flupot-pixi, so I guess it's not limited to just DOM.
